src/ModelingSystem/ModelingSystemGAMS.cpp

```diff
@@ -1629,52 +1746,84 @@ NonlinearExpressionPtr ModelingSystemGAMS::parseGamsInstructions(int codelen, /*
+                else
+                {
+                    auto expression = std::make_shared<ExpressionProduct>(
+                        std::make_shared<ExpressionConstant>(variable->lowerBound), stack.rbegin()[0]);
+                    stack.pop_back();
+                    stack.push_back(std::move(expression));
+                }
             }
             else
             {
-                auto expression = std::make_shared<ExpressionProduct>(
-                    std::make_shared<ExpressionVariable>(destination->getVariable(address)), stack.rbegin()[0]);
-                stack.pop_back();
-                stack.push_back(expression);
+                if(mainIsProd)
+                {
+                    std::dynamic_pointer_cast<ExpressionProduct>(stack.rbegin()[0])
+                        ->children.add(std::make_shared<ExpressionVariable>(variable));
+                }
+                else
+                {
+                    auto expression = std::make_shared<ExpressionProduct>(
+                        std::make_shared<ExpressionVariable>(variable), stack.rbegin()[0]);
+                    stack.pop_back();
+                    stack.push_back(std::move(expression));
+                }
             }
 
             break;
         }
 
         case nlMulI: // multiply immediate
         {
-            auto expression = std::make_shared<ExpressionProduct>(
-                std::make_shared<ExpressionConstant>(constants[address]), stack.rbegin()[0]);
-            stack.pop_back();
-            stack.push_back(expression);
+            bool mainIsProd = (stack.rbegin()[0]->getType() == E_NonlinearExpressionTypes::Product);
+
+            if(mainIsProd)
+            {
+                std::dynamic_pointer_cast<ExpressionProduct>(std::move(stack.rbegin()[0]))
+                    ->children.add(std::make_shared<ExpressionConstant>(constants[address]));
+            }
+            else
+            {
+                auto expression = std::make_shared<ExpressionProduct>(
+                    std::make_shared<ExpressionConstant>(constants[address]), std::move(stack.rbegin()[0]));
+                stack.pop_back();
+                stack.push_back(expression);
+            }
+
             break;
         }
 
         case nlMulIAdd: // multiply immediate and add
         {
             auto expressionProduct = std::make_shared<ExpressionProduct>(
-                std::make_shared<ExpressionConstant>(constants[address]), stack.rbegin()[0]);
+                std::make_shared<ExpressionConstant>(constants[address]), std::move(stack.rbegin()[0]));
             stack.pop_back();
-            stack.push_back(expressionProduct);
-            auto expressionSum = std::make_shared<ExpressionSum>(stack.rbegin()[1], stack.rbegin()[0]);
+            stack.push_back(std::move(expressionProduct));
+
+            auto expressionSum
+                = std::make_shared<ExpressionSum>(std::move(stack.rbegin()[1]), std::move(stack.rbegin()[0]));
```

Comment:
Here you do not seem to do some special handling for `stack.rbegin()[1]` being a sum.